sdk/storage/azure-storage-internal-avro/src/main/java/com/azure/storage/internal/avro/implementation/AvroParser.java

```diff
@@ -0,0 +1,112 @@
+import reactor.core.publisher.Flux;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * A class that represents a push based AvroParser that can parse avro data from a reactive stream.
+ *
+ * The parser stores the {@link AvroParserState current state}, the sync marker (parsed from the header),
+ * the file type (parsed from the header metadata), and the list of records collected so far.
+ *
+ * For each ByteBuffer that is emitted from the stream, we first add it to the state's cache, while we can make
+ * progress in parsing schema on the stack, we continue to progress in parsing schemas on the stack. When a schema
+ * identifies itself as done, we first pop the schema off the stack, then publish the result by calling the schema's
+ * result handler. The result handler will publish it to the location identified by the parent of the schema, and the
+ * parent could add another item to the stack or determine that it is done.
+ *
+ */
+public class AvroParser {
+
+    private AvroParserState state; /* State of the Avro Parser. */
+    private byte[] syncMarker; /* Sync marker read in the header. */
+    private AvroType objectType; /* Type of objects in this Avro file. */
+    private List<Object> records; /* Holds records collected so far. */
+
+    public AvroParser() {
+        this.state = new AvroParserState();
+        this.records = new ArrayList<>();
+
+        /* TODO (gapra): Make this a record schema since it essentially does the same stuff. */
+        /* Start off by adding the header schema to the stack so we can parse it. */
+        AvroHeaderSchema headerSchema = new AvroHeaderSchema(
+            objectType -> this.objectType = objectType, /* Handler to store the objectType. */
+            syncMarker -> this.syncMarker = syncMarker, /* Handler to store the sync marker. */
+            this.state,
+            this::onHeader
+        );
+        headerSchema.add();
+    }
+
+    /**
+     * Header handler.
+     *
+     * @param ignore null
+     */
+    private void onHeader(Object ignore) {
+        /* On reading the header, read a block. */
+        onBlock(ignore);
+    }
+
+    /**
+     * Block handler.
+     *
+     * @param ignore null
+     */
+    private void onBlock(Object ignore) {
+        /* On reading the block, read a header. */
+        AvroBlockSchema blockSchema = new AvroBlockSchema(
+            this.objectType,
+            o -> this.records.add(o), /* Object result handler. */
+            this.syncMarker,
+            this.state,
+            this::onBlock
+        );
+        blockSchema.add();
+    }
+
+    /**
+     * Parses as many objects as possible in the buffer.
+     *
+     * Caches buffer remaining as part of state in case the entire buffer is not used.
+     * @param buffer {@link ByteBuffer} that is part of an Avro file.
+     * @return A reactive stream of Objects found in this
+     */
+    public Flux<Object> parse(ByteBuffer buffer) {
```

Comment:
Why are we returning a flux here? Or anything for that matter? My understanding is we are pushing results through `schema.publish()`.

Reply:
schema.publish will just call the result handler for the specific schema on top of the stack (in our case, it will add it to the global list of objects in this class)

Reply:
To elaborate, if I understand correctly, this is sort of the same pattern as the UploadBufferPool. You essentially flatMap all the data from your source into calls to this method, and it'll return a bunch of empty fluxes as it's reading stuff in and then periodically emit one that has an actual item.

Reply:
What Rick is saying is right, but say a ByteBuffer were to contain enough data to emit like say 5 different Objects, it'll emit them all together (after processing as many schemas as possible).